Problem: Test test_websocket_logs_invalid_auth fail on Python 3.12.3 (#692)
aleph-im/aleph-vm: doc/operator_auth.md

 Authentication protocol for VM owner
=======================================
This custom protocol allow a user (owner of a VM) to securely authenticate to a CRN, using their Ethereum wallet.
This scheme was designed in a way that's convenient to be integrated in the console web page.

It allows the user to control their VM. e.g : stop reboot, view their log, etc…

> ## Overview

Reviewer: ```suggestion
## Motivations
This protocol ensures secure authentication between a blockchain wallet owner and an aleph.im compute node.

Private key access is typically gated by prompts requiring manual approval for each signing operation. With hardware wallets, users are prompted both by the software on their device and the hardware wallet itself.

## Overview
```

Author: not sur I understand the second paragraph. is that a llm suggestion?

Reviewer: What I mean there is that when using a client (CLI or web) with a blockchain wallet, the user usually has to sign every operation with a prompt.

For example with Metamask:

![Screenshot 2024-09-17 at 10-42-37 MetaMask](https://github.com/user-attachments/assets/6d5a0f25-297a-45f9-856b-7b8874d661fb)

When the private key is stored in a hardware wallet, it is one prompt on screen plus one on the hardware wallet.

![image](https://github.com/user-attachments/assets/8d08d08d-7674-4cad-bdac-e01aa5b01015)

This is very annoying to do for every operation on a CRN (access logs, ...).

I want to explain why this intermediate temporary key pair is created.

The client generates a JWK key pair and signs the public key with their Ethereum account. The signed public key is sent
in the `X-SignedPubKey` header. The client also signs the operation payload with the private JWK, sending it in the
`X-SignedOperation` header. The server verifies both the public key and payload signatures, ensuring the request's
integrity and authenticity. If validation fails (e.g., expired key or invalid signature), the server returns a 401
Unauthorized error.

> Support for the Solana wallet will be added soon.

Reviewer: Support for Solana wallets is planned in the near future.

> Support for the Solana wallet will be added soon.

Author: ```suggestion

Support for Solana wallets is planned in the near future.

```

}
```

This public key is signed using the Ethereum account to ensure its authenticity. The resulting signature is
combined with the public key into a payload and sent as the `X-SignedPubKey` header.

### 2. Sign Operation Payload

#### Operation Payload Format

The operation payload is a JSON object that encapsulates the details of an API request. It ensures that the request's
integrity can be verified through signing. Below are the fields included:

- **`time`**: (string, ISO 8601 format) The timestamp for when the operation is valid, including the timezone is mandatory (`Z`
  indicates UTC). This helps prevent replay attacks (capturing the packet and replying it multiple time). e.g. `"2010-12-25T17:05:55Z"`
- **`method`**: (string) The HTTP method used for the operation (e.g., `GET`, `POST`).
- **`path`**: (string) The endpoint path of the request (e.g., `/`).
- **`domain`**: (string) The domain associated with the request. This ensures the request is valid for the intended
  CRN. (e.g., `localhost`).

Example

```json
{
  "time": "2010-12-25T17:05:55Z",
  "method": "GET",
  "path": "/",
  "domain": "localhost"
}
```

It is sent serialized as a hex string.

#### Signature

> This payload is serialized, signed, and sent in the `X-SignedOperation` header to ensure the integrity and authenticity

Reviewer: How is the payload serialized ?

of the request.

* The operation payload (containing details such as time, method, path, and domain) is serialized and converted into a byte array.
* The JWK (private key) is used to sign this operation payload, ensuring its integrity. This signature is then included in the X-SignedOperation header.

###  3. Include authentication Headers
These two headers are to be added to the HTTP Request:

1. **`X-SignedPubKey` Header:**
    - This header contains the public key payload and the signature of the public key generated by the Ethereum account.

   Example:
   ```json
   {
      "payload": "<hexadecimal string of the public key payload>",
      "signature": "<Ethereum signed public key>"
   }
   ```

> 2. **`X-SignedOperation` Header:**

Reviewer: Is it normal I get text highlighted in red on GitHub ?

![image](https://github.com/user-attachments/assets/f34c916f-d87e-4246-baf1-f22c580ca12b)

Author: saw it not sure why it happend, is red in the diff but in the final version it looks correct

![image](https://github.com/user-attachments/assets/2f1edb28-d51e-4044-8f8a-597c25bea9fb)

    - This header contains the operation payload and the signature of the operation payload generated using the private
      JWK.

   Example:
   ```json
   {
      "payload": "<hexadecimal string of the operation payload>",
      "signature": "<JWK signed operation payload>"
   }
   ```

### Expiration and Validation

- The public key has an expiration date, ensuring that keys are not used indefinitely.
- Both the public key and the operation signature are validated for authenticity and integrity at the server side.
- Requests failing verification or expired keys are rejected with `401 Unauthorized` status, providing an error message
  indicating the reason.

### WebSocket Authentication Protocol

In the WebSocket variant of the authentication protocol, the client establishes a connection and authenticates through
an initial message that includes their Ethereum-signed identity, ensuring secure communication.

> The difference is that the two header are sent in one json packet, under the `auth` key

Reviewer: ```suggestion
Due to web browsers not allowing custom HTTP headers in WebSocket connections, the two header are sent in one json packet, under the `auth` key
```

            "signature": "0xea99ef5f1a10f2d103f94dce4f8650730315246e6d15cf9e5862c11adfd6482703cd1ec684a4f3dffb36ae5c4a57b08a47108fe55e3b2454e45f6e63342e0f471b"
        },
        "X-SignedOperation": {
            "payload": "7b2274696d65223a2022323031302d31322d32355431373a30353a35355a222c20226d6574686f64223a2022474554222c202270617468223a20222f222c2022646f6d61696e223a20226c6f63616c686f7374227d",
            "signature": "6f737654cd00e4d4155d387509978e7a9a4d27f5b59c9492ac1dec7b09f9aecc58c9365526bbddd6211b65f40f4956c50ab26f395f7170ce1698c11e28e25d3a"
        }
    }
}
```

If the  authentication  succeed the server will answer with
```json
{
  "status": "connected"
}
```

In case of failed auth the server will respond with await `{"status": "failed", "reason": "string describing the reason"})` and close the connexion  

> Note: Authentication via Headers are not used for the websocket transport as it is blocked by some browsers.

Reviewer: Only "some" browsers ? Which ones ?

Author: no idea, not sure it matter
